docs/training/apis-howto/deepspeed/autotuning.rst

```diff
@@ -0,0 +1,247 @@
+              outputs = model_engine(inputs)
+              loss = criterion(outputs, labels)
+              model_engine.backward(loss)
+              model_engine.step()
+
+where ``core_context`` is the :class:`~determined.core.Context` instance which was initialized with
+``determined.core.init``. The context manager requires access to both ``core_context`` and the
+current :class:`~determined.core.SearcherOperation` instance (``op``) in order to appropriately report
+results.
+
+
+HuggingFace Trainer
+===================
+
+DeepSpeed Autotune can also be used with the HuggingFace (HF) Trainer and Determined AI's
+:class:`~determined.integrations.huggingface.DetCallback` callback object.
+
+As in the above cases, a ``deepspeed_config`` field  specifying
+the relative path to the DS ``json`` config file must again be added to the
+``hyperparameters`` section of the experiment configuration. Reporting results back to the
+Determined master now requires both using the ``dsat.dsat_reporting_context`` context manager and 
+the ``DetCallback`` callback object listed above.  Additionally, because ``dsat`` performs a search
+over different batch sizes and HuggingFace expects parameters to be specified through command-line
+arguments, an additional helper is needed to create consistent HuggingFace arguments:
+:func:`~dsat.get_hf_args_with_overwrites``.
+
+The key pieces of relevant code from a HuggingFace Trainer script are below.
+
+.. code:: python
+
+  from determined.integrations.huggingface import DetCallback
+  from determined.pytorch import dsat
+  from transformers import HfArgumentParser,Trainer, TrainingArguments,
+
+  hparams = self.context.get_hparams()
+  parser = HfArgumentParser(TrainingArguments)
+  args = sys.argv[1:]
+  args = dsat.get_hf_args_with_overwrites(args, hparams)
+  training_args = parser.parse_args_into_dataclasses(args, look_for_args_file=False)
+
+  det_callback = DetCallback(core_context, ...)
+  trainer = Trainer(args=training_args, ...)
+  with dsat.dsat_reporting_context(core_context, op=det_callback.current_op):
+      train_result = trainer.train(resume_from_checkpoint=checkpoint)
+
+
+Things to note:
+
+* The ``dsat_reporting_context`` context manager shares the same initial
+  :class:`~determined.core.SearcherOperation` as the ``DetCallback`` instance through its
+  ``op=det_callback.current_op`` argument.
+
+* The entire ``train`` method of the HuggingFace trainer is now wrapped in the
+  ``dsat_reporting_context`` context manager.
+
+
+
+Advanced Options
+================
+
+The command-line entrypoint to ``dsat`` has various available options, some of them
+search-algorithm-specific.  All available options for any given search method can be found as in
+
+.. code::
+
+   python3 -m determined.pytorch.dsat asha --help
+
+and similar for the other search methods. Below, we highlight particularly important flags and
+describe the search algorithms in some more detail.
+
+
+General Options
+^^^^^^^^^^^^^^^
```

Comment:
- review of General Options up through "asha" (stopped at "asha")

**General Options**

The following options are available for every search method:

* By default, ``dsat`` launches 50 trials and runs up to 16 trials concurrently. You can modify these values using the ``--max-trials`` and ``--max-concurrent-trials`` flags. In addition, you can limit the number of trials by specifying the ``--max-slots`` flag.

* Other notable flags include:

 * ``--max-trials``: Sets the maximum total number of trials to run. The default value is 50.
 * ``--max-concurrent-trials``: Specifies the maximum total number of trials that can run concurrently. The default value is 16.
 * ``--max-slots``: Specifies the maximum total number of slots that can run concurrently. The default is ``None``, meaning there is no limit by default.
 * ``--metric``: Specifies the metric to be optimized during autotuning. The default metric is ``FLOPS-per-gpu``. Other options are ``throughput``, ``forward``, ``backward``, and ``latency``.
 * ``--run-full-experiment``: When this flag is specified, after every ``dsat`` trial is completed, a single-trial experiment will be launched using the specifications in the ``deepspeed.yaml`` file, which will be overwritten with the best-found DS configuration parameters.
 * ``--zero-stages``: By default, ``dsat`` searches over each of the stages: ``1``, ``2``, and ``3``. To limit the search to a specific subset of stages, use the ``--zero-stages`` flag followed by a space-separated list (e.g., ``--zero-stages 2 3``).